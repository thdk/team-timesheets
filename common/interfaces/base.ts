import { FieldValue, Timestamp } from "firebase/firestore";
export interface INameWithIcon {
    name: string;
    icon?: string;
    created?: Date;
    deleted?: boolean;
}

export interface INameWithIconData extends IPersistedEntity {
    name: string;
    name_insensitive?: string;
    icon?: string;
    deleted?: boolean;
}

export interface IPersistedEntity {
    created?: Timestamp;
    modified?: Timestamp;
}

export interface IFlaggedForDeletion {
    deleted: boolean;
    modified: FieldValue;
}
