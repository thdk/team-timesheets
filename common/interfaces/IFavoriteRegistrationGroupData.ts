export interface IFavoriteRegistrationGroupData  {
    name?: string;
    userId: string;
};
