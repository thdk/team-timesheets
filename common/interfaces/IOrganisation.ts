import { INameWithIcon } from "..";

export interface IDivision extends INameWithIcon {
    createdBy: string;
    id: string;
}