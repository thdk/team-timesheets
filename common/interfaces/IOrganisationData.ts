import { INameWithIconData } from "./base";

export interface IOrganisationData extends INameWithIconData {
    createdBy: string;
    id: string;
}