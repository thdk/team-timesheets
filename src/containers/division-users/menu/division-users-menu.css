.division-users-menu.mdc-menu-surface {
    width: 220px;
    top: 62px !important;
}