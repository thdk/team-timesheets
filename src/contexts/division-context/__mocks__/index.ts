export const useDivisionStore = jest.fn().mockReturnValue({
    userDivisions: [],
});
