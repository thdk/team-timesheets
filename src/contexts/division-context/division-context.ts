import { useStore } from "../store-context";

export const useDivisionStore = () => {
    const store = useStore();
    return store.divisions;
}