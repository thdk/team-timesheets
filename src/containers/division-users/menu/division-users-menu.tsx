import React, { useCallback, useState } from "react";
import { MenuSurfaceAnchor, MenuItem, MenuHTMLProps, MenuSurface } from "@rmwc/menu";
import { ListItem, List, ListDivider, ListItemGraphic, ListItemText } from "@rmwc/list";
import { observer } from "mobx-react-lite";

import { useDivisionStore } from "../../../contexts/division-context";
import { useRouterStore } from "../../../stores/router-store";
import { goToUserProfile } from "../../../internal";
import { useUserStore } from "../../../contexts/user-context";
import { AccountBadge } from "../../../components/account-badge";
import { IDivision } from "../../../../common/interfaces/IOrganisation";

import "./division-users-menu.css";

export const DivisionUsersMenu = observer(({
    children,
    ...menuProps
}: MenuHTMLProps
) => {
    const division = useDivisionStore();
    const router = useRouterStore();
    const user = useUserStore();

    const [isMenuOpen, setIsMenuOpen] = useState(false);

    const handleSelect = useCallback((selectedDivision: (IDivision & { divisionUserId: string }) | undefined) => {
        setIsMenuOpen(false);
        user.updateAuthenticatedUser({
            divisionUserId: selectedDivision?.divisionUserId,
            divisionId: selectedDivision?.id,
        });
    }, [goToUserProfile, user, division]);

    const handleAccountOnClick = useCallback(() => {
        if (division.userDivisions.length > 1) {
            setIsMenuOpen(!isMenuOpen);
        }
        else {
            goToUserProfile(router);
        }
    }, [router, division, setIsMenuOpen, goToUserProfile]);

    const displayName = division.division || user.authenticatedUser?.divisionId
        ? (
            division.division
                ? division.division.data!.name
                : ""
        )
        : (
            user.divisionUser
                ? user.divisionUser.name || "Guest"
                : ""
        );

    const email = user.divisionUser
        ? user.divisionUser.email || "unknown@timesheets.com"
        : "";

    return (
        <MenuSurfaceAnchor>
            <MenuSurface
                anchorCorner={"bottomStart"}
                open={isMenuOpen}
                onClose={() => setIsMenuOpen(false)}
                className={"division-users-menu"}
                {...menuProps}
            >
                <List>
                    {division.userDivisions.map(division => <ListItem
                        role="menuitem"
                        tabIndex={0}
                        key={division.divisionUserId}
                        onClick={() => handleSelect(division)}
                    >
                        <ListItemGraphic icon={division.icon} />
                        <ListItemText>
                            {division.name}
                        </ListItemText>
                    </ListItem>)}
                    <ListDivider />
                    <ListItem
                        onClick={() => handleSelect(undefined)}>
                        <ListItemGraphic
                            icon={"all_inclusive"}
                        />
                        <ListItemText>
                            Default
                        </ListItemText>
                    </ListItem>
                </List>
                <ListDivider />
                <List>
                    <MenuItem
                        onClick={() => {
                            setIsMenuOpen(false);
                            goToUserProfile(router);
                        }}
                    >
                        Manage divisions
                    </MenuItem>
                    <MenuItem
                        onClick={() => {
                            setIsMenuOpen(false);
                            goToUserProfile(router);
                        }}
                    >
                        Preferences
                    </MenuItem>
                </List>

            </MenuSurface>
            <AccountBadge
                email={email}
                name={displayName}
                onClick={handleAccountOnClick}
            />
        </MenuSurfaceAnchor>
    )
});
