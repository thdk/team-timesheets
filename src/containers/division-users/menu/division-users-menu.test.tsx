import React from "react";
import { useDivisionStore } from "../../../contexts/division-context";
import { DivisionUsersMenu } from "./division-users-menu";
import { render } from "@testing-library/react";
import { IDivision } from "../../../../common/interfaces/IOrganisation";


jest.mock('../../configs/use-configs');
jest.mock('../../../stores/router-store');
jest.mock('../../../contexts/division-context');
jest.mock('../../configs/use-configs');

jest.mock('../../../contexts/user-context');

describe("DivisionUsersMenu", () => {
    beforeAll(() => {
        (useDivisionStore as jest.Mock<ReturnType<typeof useDivisionStore>>).mockReturnValue(
            {
                userDivisions: [] as (IDivision & { divisionUserId: string })[],
            } as ReturnType<typeof useDivisionStore>
        );
    });

    it("should render", () => {
        const {
            asFragment,
        } = render(
            <DivisionUsersMenu />
        );

        expect(asFragment()).toMatchSnapshot();
    });
});
