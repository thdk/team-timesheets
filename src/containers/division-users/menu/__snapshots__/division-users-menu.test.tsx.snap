// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DivisionUsersMenu should render 1`] = `
<DocumentFragment>
  <div
    class="mdc-menu-surface--anchor"
  >
    <div
      class="division-users-menu mdc-menu-surface"
      style="transform-origin: left top; left: 0px; top: 0px; max-height: 736px;"
    >
      <ul
        class="mdc-list"
      >
        <li
          class="mdc-list-divider"
          role="separator"
        />
        <li
          class="mdc-list-item"
          tabindex="0"
        >
          <i
            aria-hidden="true"
            class="rmwc-icon rmwc-icon--ligature material-icons mdc-list-item__graphic"
          >
            all_inclusive
          </i>
          <span
            class="mdc-list-item__text"
          >
            Default
          </span>
        </li>
      </ul>
      <li
        class="mdc-list-divider"
        role="separator"
      />
      <ul
        class="mdc-list"
      >
        <li
          class="mdc-list-item"
          role="menuitem"
          tabindex="0"
        >
          Manage divisions
        </li>
        <li
          class="mdc-list-item"
          role="menuitem"
          tabindex="0"
        >
          Preferences
        </li>
        <li
          class="mdc-list-divider"
          role="separator"
        />
        <li
          class="mdc-list-item"
          role="menuitem"
          tabindex="0"
        >
          <i
            aria-hidden="true"
            class="rmwc-icon rmwc-icon--ligature material-icons mdc-list-item__graphic"
          >
            perm_identity
          </i>
          <span
            class="mdc-list-item__text"
          >
            Logout
          </span>
        </li>
      </ul>
    </div>
    <div
      class="account-badge"
    >
      <div
        class="account-badge__content"
      >
        <div
          class="account-badge__image"
        >
          <img
            src="https://www.gravatar.com/avatar/db2ff1aecc3bb4c910ab7289c7f920a2?d=mp&s=40"
          />
        </div>
        <div
          class="account-badge__name"
        >
          Foobar
        </div>
        <div
          class="account-badge__meta"
        >
          <i
            class="rmwc-icon rmwc-icon--ligature material-icons"
          >
            expand_more
          </i>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
